hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionRegistryRpcStubHolder.java

```diff
@@ -0,0 +1,212 @@
+          synchronized (ConnectionRegistryRpcStubHolder.this) {
+            addr2StubFuture = null;
+            if (controller.failed()) {
+              if (ConnectionUtils.isUnexpectedPreambleHeaderException(controller.getFailed())) {
+                // this means we have connected to an old server where it does not support passing
+                // cluster id through preamble connnection header, so we fallback to use null
+                // cluster id, which is the old behavior
+                LOG.debug("Failed to get connection registry info, should be an old server,"
+                  + " fallback to use null cluster id", controller.getFailed());
+                createStubsAndComplete(null, future);
+              } else {
+                LOG.debug("Failed to get connection registry info", controller.getFailed());
+                if (index == bootstrapServers.size() - 1) {
+                  future.completeExceptionally(controller.getFailed());
+                } else {
+                  // try next bootstrap server
+                  createStubs(bootstrapServers, index + 1, future);
+                }
+              }
+            } else {
+              LOG.debug("Got connection registry info: {}", resp);
+              String clusterId = resp.getClusterId();
+              createStubsAndComplete(clusterId, future);
+            }
+          }
+          getConnectionRegistryRpcClient.close();
+        }
+      });
+  }
+
+  private CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> createStubs() {
+    CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> future =
+      new CompletableFuture<>();
+    List<ServerName> bootstrapServers = new ArrayList<ServerName>(bootstrapNodes);
+    Collections.shuffle(bootstrapServers);
+    createStubs(bootstrapServers, 0, future);
+    return future;
+  }
+
+  CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> getStubs() {
+    ImmutableMap<ServerName, ClientMetaService.Interface> s = this.addr2Stub;
+    if (s != null) {
+      return CompletableFuture.completedFuture(s);
+    }
+    synchronized (this) {
+      s = this.addr2Stub;
+      if (s != null) {
+        return CompletableFuture.completedFuture(s);
+      }
+      if (addr2StubFuture != null) {
+        return addr2StubFuture;
+      }
+      addr2StubFuture = createStubs();
+      return addr2StubFuture;
+    }
+  }
+
+  void refreshStubs(IOExceptionSupplier<Collection<ServerName>> fetchEndpoints) throws IOException {
+    // There is no actual call yet so we have not initialize the rpc client and related stubs yet,
+    // give up refreshing
+    if (addr2Stub == null) {
+      LOG.debug("Skip refreshing stubs as we have not initialized rpc client yet");
+      return;
+    }
+    LOG.debug("Going to refresh stubs");
+    assert rpcClient != null;
+    addr2Stub = createStubs(rpcClient, fetchEndpoints.get());
```

Comment:
do we not need to synchronize the setting of addr2Stub? Looks like we are doing it in 2 places, here and above, and we're also reading it in a synchronized block above.

Reply:
I now see above that we are synchronizing in the callback of createStubs, but still seems like we need it here too

Reply:
No, we don't. It is the different createStubs method, where we just update the addr2Stubs field without setting up any real rpc connections. It is a bit confusing as we use the same method name. Let me see if we can change this a bit.

Reply:
Let me check if we can change the code like this.

For future vs. IOExce[tionSupplier, the intention here is that we do not need asynchronous call here, we are in a background thread, we just fetch the result in a blocking way and update the addr2Stub, so we do not need a CompletableFuture here, the only thing is that we use an async client in the caller method so actually the implementation is a future.get.

Reply:
We do not need to synchronize here.

The synchronization is for protecting the initialization of addr2Stub, and after initialization, addr2Stub will never be null, so we will never enter the synchronization block in getStubs method any more. And since addr2Stub is volatile, in refreshStubs, we just update the reference to addr2Stub at once, without changing its content, it is safe to not do any synchronization. Is this clear enough? Or we could introduce a initLock and synchronized on this object in getStubs method to make the logic more clear?

Reply:
Ok thanks, that makes sense as is